File: BOJ_백준/16922 로마 숫자 만들기 (dfs).cpp
#include <iostream>

using namespace std;

int N, ans;
int roma[4] = { 1, 5, 10, 50 };
bool visited[50 * 20 + 1];

void dfs(int now, int cnt, int num) {
	if (cnt == N) {
		if (visited[num] == false) {
			visited[num] = true;
			ans++;
		}
		return;
	}

	for (int i = now; i < 4; i++) {
		dfs(i, cnt + 1, num + roma[i]);
	}
}

int main() {
	ios_base::sync_with_stdio(false);
	cin.tie(NULL);
	
	cin >> N;
	dfs(0, 0, 0);
	cout << ans;
	return 0;
}

File: BOJ_백준/16922 로마 숫자 만들기 (dfs).py
import sys
sys.setrecursionlimit(10**6)

N = int(input())
visited = [False] * (50 * 20 + 1)
roma = [1, 5, 10, 50]
ans = 0
def dfs(now: int, cnt: int, num: int) :
    global ans
    if cnt == N :
        if not visited[num] :
            visited[num] = True
            ans += 1
        return

    for i in range(now, 4) :
        dfs(i, cnt + 1, num + roma[i])


dfs(0, 0, 0)
print(ans)
